firmware/application/ui_external_items_menu_loader.cpp

```diff
@@ -0,0 +1,137 @@
+#include "ui_external_items_menu_loader.hpp"
+
+#include "sd_card.hpp"
+
+namespace ui {
+
+std::vector<Bitmap*> ExternalItemsMenuLoader::bitmaps;
+
+std::vector<GridItem> ExternalItemsMenuLoader::load_external_items(app_location_t app_location, NavigationView& nav) {
+    for (const auto& bitmap : bitmaps) {
+        delete bitmap->data;
+        delete bitmap;
+    }
+    bitmaps.clear();
+
+    std::vector<GridItem> exteral_apps;
+
+    if (sd_card::status() != sd_card::Status::Mounted)
+        return exteral_apps;
+
+    for (const auto& entry : std::filesystem::directory_iterator(u"APPS", u"*.ppma")) {
+        FIL ppma_file;
+        auto filePath = u"/APPS/" + entry.path().native();
+        if (f_open(&ppma_file, reinterpret_cast<const TCHAR*>(filePath.c_str()), FA_READ) != FR_OK)
+            continue;
+
+        size_t bytes_read;
+        application_information_t application_information = {};
+        if (f_read(&ppma_file, &application_information, sizeof(application_information_t), &bytes_read) != FR_OK) continue;
+
+        if (application_information.menu_location != app_location)
+            continue;
+
+        bool versionMatches = VERSION_MD5 == application_information.app_version;
+
+        GridItem gridItem = {};
+        gridItem.text = reinterpret_cast<char*>(&application_information.app_name[0]);
+
+        if (versionMatches) {
+            gridItem.color = Color((uint16_t)application_information.icon_color);
+
+            auto bitmapData = new uint8_t[32];
+            Bitmap* bitmap = new Bitmap{
+                {16, 16},
+                bitmapData};
+
+            memcpy(bitmapData, &application_information.bitmap_data[0], 32);
+            bitmaps.push_back(bitmap);
+
+            gridItem.bitmap = bitmap;
+
+            gridItem.on_select = [&nav, app_location, filePath]() {
+                auto filename = reinterpret_cast<const TCHAR*>(filePath.c_str());
+                run_external_app(nav, filename);
+            };
+        } else {
+            gridItem.color = Color::light_grey();
+
+            gridItem.bitmap = &bitmap_sd_card_error;
+
+            gridItem.on_select = [&nav, app_location, filePath]() {
+                nav.display_modal("Error", "The .ppma file in your APPS\nfolder is outdated. Please\nupdate your SD Card content.");
```

Comment:
Might indicate which file you bailed on considering this is part of discovery.

Reply:
This is shown after you clicked on one specific app. and the probability that all apps are outdated is very high. I'll leave it like that